core/src/kmx/kmx_plus.cpp

```diff
@@ -953,6 +953,89 @@ COMP_KMXPLUS_LIST_Helper::getIndex(KMX_DWORD i) const {
   return indices + i;
 }
 
+
+// USET
+
+bool
+COMP_KMXPLUS_USET::valid(KMX_DWORD _kmn_unused(length)) const {
+  if (header.size < sizeof(*this)
+      + (usetCount  * sizeof(COMP_KMXPLUS_USET_USET))
+      + (rangeCount * sizeof(COMP_KMXPLUS_USET_RANGE))) {
+    DebugLog("header.size < expected size");
+    assert(false);
+    return false;
+  }
+  return true;
+}
+
+COMP_KMXPLUS_USET_Helper::COMP_KMXPLUS_USET_Helper() : uset(nullptr), is_valid(false), usets(nullptr), ranges(nullptr) {
+}
+
+bool
+COMP_KMXPLUS_USET_Helper::setUset(const COMP_KMXPLUS_USET *newUset) {
+  DebugLog("validating newUset=%p", newUset);
+  is_valid = true;
+  if (newUset == nullptr) {
+    // null = invalid
+    is_valid = false;
+    // No assert here: just a missing layer
+    return false;
+  }
+  uset = newUset;
+  const uint8_t *rawdata = reinterpret_cast<const uint8_t *>(newUset);
+  rawdata += LDML_LENGTH_USET;  // skip past non-dynamic portion
+  // usets
+  if (uset->usetCount > 0) {
+    usets = reinterpret_cast<const COMP_KMXPLUS_USET_USET *>(rawdata);
+  } else {
+    usets = nullptr;
+    // not invalid, just empty.
+  }
+  rawdata += sizeof(COMP_KMXPLUS_USET_USET) * uset->usetCount;
+  // entries
+  if (uset->rangeCount > 0) {
+    ranges = reinterpret_cast<const COMP_KMXPLUS_USET_RANGE *>(rawdata);
+  } else {
+    ranges = nullptr;
+  }
```

Comment:
Are we guaranteed to have checked validity before calling `setUset`? Otherwise, if these values `uset->usetCount` and `uset->rangeCount` are out of range, this could get access violations.

Reply:
Guaranteed, `section_from_sect()` at 

https://github.com/keymanapp/keyman/pull/9197/files#diff-95985cb44c34779804524113d19f84c84c0df3809790667a95bd3592e3a9e7cbR1087 

checks validity or returns nullptr.  So `uset` would be null.  The key is that that function has the *original ending size* of the kmx+ region, which isn't propagated further.

Same for every section and helper.

At this point:

- the _Helper things should be children of a templatized base class. 
- `COMP_KMXPLUS_*::valid(size)` functions should have `size` parameter removed, because it's used there, the size is checked in the header against the end of the region, and it's not needed further.